apis/rbac/v1alpha1/types.go

```diff
@@ -0,0 +1,302 @@
+	PrincipalIDRef *xpv1.Reference `json:"principalIDRef,omitempty"`
+
+	// PrincipalIDRef - Select a reference to the Principal id.
+	// +immutable
+	PrincipalIDSelector *xpv1.Selector `json:"principalIDSelector,omitempty"`
+
+	// GroupIDRef - A reference to the Group id.
+	// +optional
+	// +immutable
+	GroupIDRef *xpv1.Reference `json:"groupIDRef,omitempty"`
+
+	// GroupIDSelector - Select a reference to the Group id.
+	// +immutable
+	GroupIDSelector *xpv1.Selector `json:"groupIDSelector,omitempty"`
+
+	// RoleID - The role definition ID.
+	// +optional
+	// +kubebuilder:validation:Required
+	RoleID string `json:"roleID,omitempty"`
+
+	// RoleIDRef - A reference to the Role id.
+	// +optional
+	// +immutable
+	RoleIDRef *xpv1.Reference `json:"roleIDRef,omitempty"`
+
+	// RoleIDSelector - Select a reference to the Role id.
+	// +immutable
+	RoleIDSelector *xpv1.Selector `json:"roleIDSelector,omitempty"`
+
+	// Scope - The role assignment scope.
+	// +immutable
+	// +kubebuilder:validation:Required
+	Scope string `json:"scope"`
+}
+
+// An RoleAssignmentStatus represents the observed state of an RoleAssignment.
+type RoleAssignmentStatus struct {
+	xpv1.ResourceStatus `json:",inline"`
+}
+
+// +kubebuilder:object:root=true
+
+// A RoleAssignment is a managed resource that represents an RoleAssignment.
+// +kubebuilder:printcolumn:name="READY",type="string",JSONPath=".status.conditions[?(@.type=='Ready')].status"
+// +kubebuilder:printcolumn:name="SYNCED",type="string",JSONPath=".status.conditions[?(@.type=='Synced')].status"
+// +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp"
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope=Cluster,categories={crossplane,managed,azure}
+type RoleAssignment struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   RoleAssignmentSpec   `json:"spec"`
+	Status RoleAssignmentStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// RoleAssignmentList contains a list of RoleAssignment items
+type RoleAssignmentList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []RoleAssignment `json:"items"`
+}
+
+// Role and ADGroup are only CRDs and has no reconcile controller
+
+// An RoleSpec defines the desired state of an Role.
+type RoleSpec struct {
+	xpv1.ResourceSpec `json:",inline"`
+
+	// RoleID - The role definition ID.
+	// +kubebuilder:validation:Required
+	RoleID string `json:"roleID"`
+}
+
+// An RoleStatus represents the observed state of an Role.
+type RoleStatus struct {
+	xpv1.ResourceStatus `json:",inline"`
+}
+
+// +kubebuilder:object:root=true
+
+// A Role is a resource that represents an Role.
+// +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp"
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope=Cluster,categories={crossplane,managed,azure}
+type Role struct {
```

Comment:
My understanding is `Role` is not a managed resource, i.e., we do not have the an `ExternalConnecter` and an `ExternalClient` for `Role`s. Thus, we should not treat it as a managed resource and it should not have a CRD. For the currently defined referencer and selector fields for it, you may just leave the ID field. If there is an actual Azure resource corresponding to roles, then in a separate PR we can introduce the `Role` managed resource that actually manipulates Azure role objects.